src/els_work_done_progress.erl

```diff
@@ -0,0 +1,93 @@
+%%==============================================================================
+%% Protocol support for workDoneProgress
+%%==============================================================================
+-module(els_work_done_progress).
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% Types
+%%==============================================================================
+-type percentage() :: 0..100.
+-type value_begin() :: #{ kind := 'begin'
+                        , title := binary()
+                        , cancellable => boolean()
+                        , message => binary()
+                        , percentage => percentage()
+                        }.
+-type value_report() :: #{ kind := 'report'
+                         , cancellable => boolean()
+                         , message => binary()
+                         , percentage => percentage()
+                         }.
+-type value_end() :: #{ kind := 'end'
+                      , message => binary()
+                      }.
+-type value() :: value_begin()
+               | value_report()
+               | value_end().
+
+-export_type([ value_begin/0
+             , value_report/0
+             , value_end/0
+             , value/0
+             ]).
+
+%%==============================================================================
+%% Exports
+%%==============================================================================
+
+-export([ is_supported/0
+        , send_create_request/0
+        , value_begin/3
+        , value_report/2
+        , value_end/1
+        ]).
+
+%%==============================================================================
+%% API
+%%==============================================================================
+
+-spec is_supported() -> boolean().
+is_supported() ->
+  case els_config:get(capabilities) of
+    #{<<"window">> := #{<<"workDoneProgress">> := WorkDoneProgress }}
+      when is_boolean(WorkDoneProgress) ->
+      WorkDoneProgress;
+    _ ->
+      false
+  end.
+
+-spec send_create_request() -> els_progress:token().
+send_create_request() ->
+  Token = els_progress:token(),
+  Method = <<"window/workDoneProgress/create">>,
+  Params = #{token => Token},
+  ok = els_server:send_request(Method, Params),
+  Token.
+
+-spec value_begin(binary(), binary(), percentage()) -> value_begin().
+value_begin(Title, Message, Percentage) ->
+  #{ kind        => 'begin'
```

Comment:
Shouldn't these `kind` values be added to `erlang_ls.hrl` instead?

Reply:
Some of the _types_ have very similar names (e.g. `kind`) and to avoid collisions, we end up with very long prefixed type names (e.g. `work_done_progress_begin`). My suggestion is that we slowly move away from a centralized header file with all type definitions and specify _local_ types, exporting them when necessary. For example, look at how the _progress token_ is exported in this PR. WDYT?

Reply:
I agree that we don't want long macro names, but now we will have a mixture of both.

Reply:
I know. I just didn't want to make a big switch. Something else we should decide on and add to the guidelines.